src/main/java/org/minions/devfund/marylin/Squarelotron.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    /**
+     * Method to duplicate a grid.
+     *
+     * @param original original grid.
+     * @return returns the new grid.
+     */
+    private int[][] duplicateGrid(final int[][] original) {
+        int[][] backupGrid = new int[size][size];
+        for (int i = 0; i < size; i++) {
+            for (int j = 0; j < size; j++) {
+                backupGrid[i][j] = original[i][j];
+            }
+        }
+        return backupGrid;
+    }
+
+    /**
+     * Upside down flip the complete matrix.
+     *
+     * @param matrix grid to be upsideDownFlipCompleteMatrix.
+     * @return return the upsideDownFlipCompleteMatrix grid.
+     */
+    int[][] upsideDownFlipCompleteMatrix(final int[][] matrix) {
+        int[][] resultGrid = duplicateGrid(matrix);
+        int temp = 0;
+        for (int i = size - 1; i >= size / 2; i--) {
+            for (int j = 0; j < size; j++) {
+                resultGrid[i][j] = matrix[temp][j];
+                resultGrid[temp][j] = matrix[i][j];
+            }
+            temp++;
+        }
+        return resultGrid;
+    }
+
+    /**
+     * Upside down flip the internal matrix.
+     *
+     * @param matrix grid to be upsideDownFlipInternalMatrix.
+     * @return return the upsideDownFlipInternalMatrix grid.
+     */
+    int[][] upsideDownFlipInternalMatrix(final int[][] matrix) {
+        int[][] resultGrid = duplicateGrid(matrix);
+        int temp = 1;
+        for (int i = size - 2; i >= size / 2; i--) {
+            for (int j = 1; j < size - 1; j++) {
+                resultGrid[i][j] = matrix[temp][j];
+                resultGrid[temp][j] = matrix[i][j];
+            }
+            temp++;
+        }
+        return resultGrid;
+    }
+
+
+    /**
+     * Upside Down Flip given a ring.
+     *
+     * @param ring ring of the grid.
+     * @return the Squarelotron Upside Down Flip.
+     */
+    Squarelotron upsideDownFlip(int ring) {
+
+        //Returns the same squarelotron if the ring number is equal to the number of rings of a matrix n x n.
+        if (Math.abs(size) % 2 == 1 && ring == this.numberOfRings) {
+            return this;
+        }
+
+        int[][] resultGrid = upsideDownFlipHelper(ring);
+        return new Squarelotron(resultGrid);
+    }
+
+    /**
+     * upsideDownFlipHelper.
+     *
+     * @param ring ring number.
+     * @return a grid.
+     */
+    private int[][] upsideDownFlipHelper(int ring) {
+        int[][] resultGrid;
+        //If the ring is the internal rings.
+        if (ring != 1) {
+            resultGrid = upsideDownFlipInternalMatrix(this.squarelotron);
+        } else {
+            //complete Upside Down Flip.
+            resultGrid = upsideDownFlipCompleteMatrix(this.squarelotron);
+
+            // Restore values of the internal rings if there are.
+            for (int i = 1; i < size - 1; i++) {
+                for (int j = 1; j < size - 1; j++) {
+                    resultGrid[i][j] = this.squarelotron[i][j];
```

Comment:
You could move the for loop outside the if-else condition.
```java
// Restore values of the internal rings if there are.
for (int i = ring; i < size - ring; i++) {
    for (int j = ring; j < size - ring; j++) {
        resultGrid[i][j] = this.squarelotron[i][j];
    }
}
```

Reply:
This loop should be only in the else condition. the logic of if condition does not need this for loop